File: cr.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "lincr.h"

char out_name[256]; 

int main(int argc, char *argv[])
{
    int i=0, n;
    FILE *fin = NULL, *fout = NULL;

    double *dat = NULL, *wvl = NULL, *cr = NULL, *cc = NULL;

    int verbose = 0;

    if (argc > 0 && argc <= 3) /* parse an existing file */
    {
	fin = fopen(argv[1],"r");
	if (fin==NULL)
	{
	  fprintf(stderr,"error: cannot open file %s\n",argv[1]);
	  exit(1);
	}
	fscanf(fin,"%d\n",&n);

	dat = (double*)malloc(n*sizeof(double));
	wvl = (double*)malloc(n*sizeof(double));
	cr  = (double*)malloc(n*sizeof(double));
	cc  = (double*)malloc(n*sizeof(double));

	while(i<n)
	{
	    fscanf(fin,"%lf %lf\n",&(wvl[i]),&(dat[i]));
	    i++;
	}
	
	fclose(fin);

	if (argc == 3)
	{
	  sprintf(out_name,"%s",argv[2]);
	}
	else
	{
	  sprintf(out_name,"crdat.txt");
	}
	fout = fopen(out_name,"w");
	if (fout==NULL)
	{
	  fprintf(stderr,"error: cannot open file %s\n",out_name);
	  exit(1);
	}
    }
    else
    {
	fprintf(stderr,"Usage: %s infile [outfile]\n", argv[0]);
	return -1;
    }

    /* compute continuum-removed representation */
    removeContinuum(n,wvl,dat,0,verbose,cr);
    /* compute continuum-curve representation */
    removeContinuum(n,wvl,dat,1,verbose,cc);

    fprintf(fout,"%d\n", n);
    for (i=0; i<n; i++)
    {
	fprintf(fout,"%lf %lf %lf %lf\n",
		wvl[i],dat[i],cr[i],cc[i]); 
    }
    fclose(fout);
    printf("CR complete, %s written\n", out_name);

    free(wvl);
    free(dat);
    free(cr);
    free(cc);
    exit(0);
}

File: cr.py
import getopt, sys
import numpy as np

from lincr import removeContinuumInplace

def main(argv=None):
    class usage(Exception):
    	def __init__(self, msg):
	    self.msg = msg

    if argv is None:
        argv = sys.argv
    try:
        try:
            longopts   = ['verbose','plot',]	  
	    longoptsp  = ['infile','outfile']   
            shortopts  = ''.join([o[0] for o in longopts])
	    shortopts += ''.join([o[0]+':' for o in longoptsp])
            opts, args = getopt.getopt(argv[1:], shortopts, longopts)
                                       
        except getopt.error, msg:
            raise usage(msg)

        infile,outfile = '','crdat.txt'
	verbose,doplot = False,False
	for opt, val in opts:
	    if opt in ('--verbose','-v'):
	        verbose=True
	    elif opt in ('--plot','-p'):
	        doplot=True
	    if opt in ('--infile','-i'):
	        infile=val
	    if opt in ('--outfile','-o'):
	        outfile=val
                
            	          
    except usage, err:
        print >>sys.stderr, err.msg
        return 2	

    
    [wvl,cidat] = np.loadtxt(infile,skiprows=1).T
    crdat,ccdat = cidat.copy(),cidat.copy()
    removeContinuumInplace(wvl,crdat,0,verbose)
    removeContinuumInplace(wvl,ccdat,1,verbose)

    np.savetxt(outfile,np.c_[wvl,cidat,crdat,ccdat])
    print "CR complete, %s written"%outfile
    if doplot: 
        import pylab as pl
        pl.figure()
        pl.suptitle('Input file: '+infile)
        pl.subplot(211)
        pl.plot(wvl,cidat,label='Continuum Intact Spectrum')
        pl.plot(wvl,ccdat,label='Continuum Curve')
        pl.legend(loc='best')
       
        pl.subplot(212)
        pl.axhline(1.0,color='k')
        pl.axhline(0.0,color='k')
        pl.plot(wvl,crdat,label='Continuum Removed Spectrum')
        pl.ylim(-0.1,1.1)
        pl.xlabel('wavelength')
        pl.legend(loc='best')
       
        pl.show()


if __name__ == '__main__':
    sys.exit(main())




